### ISSUE TITLE
mutation with nested @id reverse field leads to a duplicate XID error

### ISSUE THREAD

**Issue by dtrckd** [dgraph-io/dgraph#5753]
### What version of Dgraph are you using?

tried with
*  v20.07.0-beta.Jun22 
* v20.03.3

### Have you tried reproducing the issue with the latest release?

yes

### What is the hardware spec (RAM, OS)?

Debian 4.19.118-2 x86_64 GNU/Linux

### Steps to reproduce the issue (command/config used to run Dgraph).

Schema of the graph I am using : 
```
type Node {
  id: ID!
  nameid: String! @id
  parent: Node 
  children: [Node!] @hasInverse(field: parent)
}
```

### Expected behaviour and actual result.

When trying to add a node with a mutation like this: 
```
addNode(input: {nameid: "myNewNode", parent: {nameid: "rootNode"}}) {
  node { nameid }
}
```
I get the following error:
```
couldn't rewrite mutation addNode because failed to rewrite mutation payload because duplicate XID found: rootNode
```

In previous version of dgraph, it was working doing like this.

Note, that if I add the `id` of the "rootNode" instead of the nameid field (or by replacing it), it works fine.

**Comment by pawanrawal** [comment]
Hey @dtrckd, does a node with `@id` `rootNode` already exists inside your database? I tried to reproduce this on master (same as 20.0.7 beta) with the following steps but couldn't.

### mutate

```
mutation {
    addNode(input: [{nameid: "myNewNode", parent: {nameid: "rootNode"}}]) {
      node { nameid }
    }
}
```

The above mutation succeeds. 


### query

```
query {
  queryNode {
    nameid
    parent {
      nameid
    }
    children {
      nameid
    }
  }
}
```

responds with 

```
{
  "data": {
    "queryNode": [
      {
        "nameid": "rootNode",
        "parent": null,
        "children": [
          {
            "nameid": "myNewNode"
          }
        ]
      },
      {
        "nameid": "myNewNode",
        "parent": {
          "nameid": "rootNode"
        },
        "children": []
      }
    ]
  }
}
```

and then the final mutate (same as step 1 above) produces this error

```
{
  "errors": [
    {
      "message": "couldn't rewrite query for mutation addNode because id myNewNode already exists for type Node"
    }
  ],
  "data": {
    "addNode": {
      "node": []
    }
  },
  "extensions": {
    "touched_uids": 12
  }
}
```

**Comment by dtrckd** [comment]
Hi @pawanrawal .
Thanks for your reply.

Yes the `rooNode` with "@id" already exists, this is all the point. When a add a new node I should be able to to reference the parent either with the `id` field, or the `@nameid` (which has the @id directive) (because I add node like this, and not by pushing it in the parent children.)

Does it makes sense ?

**Comment by pawanrawal** [comment]
Hey @dtrckd , yeah you are right you should be able to reference the parent node (`rootNode`) using the `nameid` field while adding a new node. I was able to do it. 

I performed the following as my second mutation and it worked without any error.

```
mutation {
    addNode(input: [{nameid: "mySecondNewNode", parent: {nameid: "rootNode"}}]) {
      node { nameid }
    }
}
```

Could you please share the full steps that you followed to reproduce this? One more thing I noticed is that the mutation in your example has incorrect syntax. Is it the full mutation that you used? The argument for `input` should be a list and not an object.

**Comment by dtrckd** [comment]
Hi @pawanrawal,

Yes, indeed the mutation in my example was not the full mutation I used (I simplified for the example so the syntax typo).

However, by inspecting a bit more my query I figured out that the XID error was caused because there was a non @id field in the parent reference like this 

```
mutation {
    addNode(input: [{nameid: "mySecondNewNode", parent: {nameid: "rootNode", someField: "foo"}}]) {
      node { nameid }
    }
}
```

When I removed it, the query run successfully.
Otherwise, I think it 's a normal behavior to throw an error since a referenced object shouldn't be altered like this, hence I close the issue.

Many thanks.
